pylearn2/utils/iteration.py

```diff
@@ -816,11 +816,21 @@ def __init__(self, dataset, subset_iterator, data_specs=None,
 
                     def fn(batch, dspace=dspace, sp=sp):
                         try:
-                              return dspace.np_format_as(batch, sp)
-                        except ValueError as e:
-                            msg = str(e) + '\nMake sure that the model and '\
-                                           'dataset have been initialized with '\
-                                           'correct values.'
+                            return dspace.np_format_as(batch, sp)
+                        except ValueError:
+                            msg = (
+                                "\nPylearn2 expects the elements of the "
+                                "dataset (e.g. X and y) to be a matrix. If "
+                                "your data is vector shaped you should "
+                                "convert it into a one-column 2D matrix i.e. "
+                                "its shape should be [num_samples, 1]. You "
+                                "should also specify the y_labels parameter "
+                                "so that pylearn2 will use an IndexSpace "
+                                "instead of a VectorSpace. This will ensure "
+                                "an automatic conversion of the vector into a "
+                                "one-hot matrix when needed"
+                            )
+                            # raise, preserving also the original message
                             reraise_as(ValueError(msg))
```

Comment:
It is nice to have a more precise error messge, but please keep the original message (`e`), and do not add the new message in all cases: there are many other reasons a ValueError could be raised at that point.
For instance, the index for a class could be out of bound, the user could try to convert an RGB image into grayscale, etc. In those cases, the new error message would not make sense.

Actually, I think the best place to put such an error message would be in IndexSpace itself, in the `np_validate` method: there, you know exactly which conversion is attempted.